src/main/java/edu/berkeley/cs186/database/index/BPlusTree.java

```diff
@@ -420,20 +439,42 @@ private void typecheck(DataBox key) {
 
     // Iterator ////////////////////////////////////////////////////////////////
     private class BPlusTreeIterator implements Iterator<RecordId> {
-        // TODO(proj2): Add whatever fields and constructors you want here.
+        // TODO(proj2): Add whatever fields and constructors you want here. Done
+        LeafNode node;
+        Iterator<RecordId> iterator;
+        DataBox key;
+
+        BPlusTreeIterator(LeafNode node) {
+            this.node = node;
+            this.iterator = this.node.scanAll();
+        }
+
+        BPlusTreeIterator(LeafNode node, DataBox key) {
+            this.node = node;
+            this.key = key;
+            this.iterator = node.scanGreaterEqual(key);
+        }
 
         @Override
         public boolean hasNext() {
-            // TODO(proj2): implement
+            // TODO(proj2): implement Done
+            if (iterator.hasNext()) return true;
+
+            // next node
+            Optional<LeafNode> rightSibling = node.getRightSibling();
+            if (!rightSibling.isPresent()) return false;
 
-            return false;
+            this.node = rightSibling.get();
+            this.iterator = key != null ? node.scanGreaterEqual(key) : node.scanAll();
```

Comment:
코드 가독성을 위해서 삼항 연산자를 안 쓰는 방법도 좋을 것 같습니다!
```suggestion
            if (key != null) {
            this.iterator = node.scanGreaterEqual(key);
            } else {
            this.iterator = node.scanAll();
            }

```

Reply:
JK 리뷰에서도 봤던 내용같은데 , 삼항 연산자를 다들 선호하지 않으시나보네요 ...

Reply:
제이케이 제이든... 비슷한가 보네요

Reply:
제이케이든 ㅋㅋㅋ 🤣